rfc/2022/2022-06-15 - Keycloak configuration managed by Terraform.md

```diff
@@ -0,0 +1,111 @@
+2. We will store state files generated by Terraform terraform.tfstate remotely maintains the state of infrastructure and configuration. 
+3. After validating the execution plan when a pull request is approved, the process will trigger the terraform apply  command to execute proposed actions in the execution plan and apply changes in Keycloak.
+
+![Keycloak - Terraform architecture](images/kc-tf-arch.png)
+
+Below are a few examples of Terraform configuration language (HCL) and how Keycloak configurations are maintained in config files.
+
+Example of connecting Terraform to Keycloak server using the client credentials:
+
+```hcl
+provider "keycloak" {
+  client_id     = "terraform"
+  client_secret = "884e0f95-0f42-4a63-9b1f-94274655669e"
+  url           = "http://localhost:8080"
+}
+```
+
+Example of creating a client (clients.tf):
+```hcl
+resource "keycloak_realm" "realm" {
+  realm   = "my-realm"
+  enabled = true
+}
+
+resource "keycloak_openid_client" "openid_client" {
+  realm_id            = keycloak_realm.realm.id
+  client_id           = "test-client"
+
+  name                = "test client"
+  enabled             = true
+
+  access_type         = "CONFIDENTIAL"
+  valid_redirect_uris = [
+    "http://localhost:8080/openid-callback"
+  ]
+
+  login_theme = "keycloak"
+
+  extra_config = {
+    "key1" = "value1"
+    "key2" = "value2"
+  }
+}
+```
+
+Example of creating user permissions in Keycloak (permissions.tf):
+```hcl
+resource "keycloak_users_permissions" "users_permissions" {
+  realm_id = keycloak_realm.realm.id
+
+  view_scope {
+    policies          = [
+      keycloak_openid_client_user_policy.test.id
+    ]
+    description       = "description"
+    decision_strategy = "UNANIMOUS"
+  }
+
+  manage_scope {
+    policies          = [
+      keycloak_openid_client_user_policy.test.id
+    ]
+    description       = "description"
+    decision_strategy = "UNANIMOUS"
+  }
+
+  map_roles_scope {
+    policies          = [
+      keycloak_openid_client_user_policy.test.id
+    ]
+    description       = "description"
+    decision_strategy = "UNANIMOUS"
+  }
+}
+```
+## Risks
+A risk of using terraform is if the developer misconfigures terraform modules which could lead to deploying resources with security vulnerabilities inadvertently. Prevention of misconfiguration must be a significant part of the development. This issue could be avoided if there is constant and increased security scanning of Terraform modules and developers upgraded Terraform’s latest module versions with security patches.
+
+## Alternatives
+Like other tools, there are several other alternatives to terraform implementing infrastructure as code (IaC) with tools like CloudFormation (an AWS service) and Ansible.
```

Comment:
What are the benefits and risks of these alternatives? Which would you suggest as the "best" alternative?

Reply:
I think the best alternative is CloudFormation if using AWS cloud for resources. The most significant benefit of CloudFormation is the service integration with AWS cloud services. and the downside is it is not cloud-agnostic.

The benefit of the Ansible is that it is very simple to learn. it's written in python language and the playbooks are written in YAML.

Reply:
@iriteshp can we try to incorporate this comment & response into the RFC to keep it up-to-date? We also talked about adding using a K8s manifest file to manage config, but I know you're working on adding that in already.